src/Lecture2.hs

```diff
@@ -261,7 +285,21 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit literal) = Right literal
+eval vars (Var variable) =
+  let var = lookup variable vars
+  in case var of
+    Nothing -> Left (VariableNotFound variable)
+    Just a  -> Right a
+eval vars (Add a b) =
+  let first = eval vars a
+  in case first of
+    Left e  -> Left e
+    Right f ->
+      let second = eval vars b
+      in case second of
+        Left e -> Left e
+        Right s -> Right (f + s)
```

Comment:
That's a great pattern matching! 👏🏻 
Once you learn about Monads, you'll be able to reduce this case to a single line ✂️